apps/desktop/src/autofill/services/desktop-fido2-user-interface.service.ts

```diff
@@ -71,36 +88,90 @@ export class DesktopFido2UserInterfaceSession implements Fido2UserInterfaceSessi
     private cipherService: CipherService,
     private accountService: AccountService,
     private logService: LogService,
-    private messagingService: MessagingService,
     private router: Router,
     private desktopSettingsService: DesktopSettingsService,
+    private windowObject: NativeWindowObject,
   ) {}
 
   private confirmCredentialSubject = new Subject<boolean>();
   private createdCipher: Cipher;
 
+  private availableCipherIdsSubject = new BehaviorSubject<string[]>(null);
+  /**
+   * Observable that emits available cipher IDs once they're confirmed by the UI
+   */
+  get availableCipherIds$(): Observable<string[]> {
+    return this.availableCipherIdsSubject.pipe(
+      filter((ids) => ids != null),
+      take(1),
+    );
+  }
+
+  private chosenCipherSubject = new Subject<string>();
+
   // Method implementation
-  async pickCredential(
-    params: PickCredentialParams,
-  ): Promise<{ cipherId: string; userVerified: boolean }> {
-    this.logService.warning("pickCredential desktop function", params);
+  async pickCredential({
+    cipherIds,
+    userVerification,
+    assumeUserPresence,
+    masterPasswordRepromptRequired,
+  }: PickCredentialParams): Promise<{ cipherId: string; userVerified: boolean }> {
+    this.logService.warning("pickCredential desktop function", {
+      cipherIds,
+      userVerification,
+      assumeUserPresence,
+      masterPasswordRepromptRequired,
+    });
 
     try {
-      await this.showUi();
+      // Check if we can return the credential without user interaction
+      // TODO: Assume user presence is undefined
+      if (cipherIds.length === 1 && !masterPasswordRepromptRequired) {
+        this.logService.debug(
+          "shortcut - Assuming user presence and returning cipherId",
+          cipherIds[0],
+        );
+        return { cipherId: cipherIds[0], userVerified: userVerification };
+      }
+
+      this.logService.debug("Could not shortcut, showing UI");
+
+      // make the cipherIds available to the UI.
+      // Not sure if the UI also need to know about masterPasswordRepromptRequired -- probably not, otherwise we can send all of the params.
+      this.availableCipherIdsSubject.next(cipherIds);
```

Comment:
No the UI needs to check this itself because it depends on which credential the user picks
```suggestion
      this.availableCipherIdsSubject.next(cipherIds);
```
See: https://github.com/bitwarden/clients/blob/cb028eadb5fb17263a4a1c0c0ffda6b23caeb0b5/apps/browser/src/autofill/popup/fido2/fido2.component.ts#L456-L460